runtime/debugger.c

```diff
@@ -141,6 +141,8 @@ static void open_connection(void)
 #endif
   dbg_in = caml_open_descriptor_in(dbg_socket);
   dbg_out = caml_open_descriptor_out(dbg_socket);
+  Lock (dbg_in);
+  Lock (dbg_out);
```

Comment:
The point is to ensure that the channel is always locked when doing I/O. This is needed because the I/O functions now temporarily unlock the channel when handling a signal. Unlocking a channel which is not locked is bad.

Reply:
Ok, but (un)locking without the thread library is  no-op. So I would prefer simply checking here that the thread library is not loaded.

Reply:
I like the "declaration of intent" to lock even if this is a noop because it makes the code more readable, but I agree that a comment clarifying with @jhjourdan's point is desirable.

Reply:
You're right that this would work, but it seems unnecessarily fragile. The comments in io.h now state that the channel should be locked while doing any I/O, so why not follow that advice here?

Reply:
Ok for letting the locking instructions here. That said, instead of a bare comment, I would prefer checking whether the systhread library and raise a fatal error if so. It is actually unsafe to use both features at the same time.

Reply:
The reason I am somewhat disturbed about these locking instructions is that they do not seem to protect anything, since any OCaml code can be executed in the critical section. So we are left choosing between two ways of writing fragile code:
  1- Code that would break if locking suddenly stops being a no-op in single-threaded mode
  2- Code that would break if somehow the debugger code was executed in a re-entrant way.

Reply:
Another issue is that in all other places, only one channel is locked at a time. This is important for the mechanism that unlocks channels when an exception arises (cf Unlock_exn), which only remembers one channel. This is not a bug here as long as they are no-ops, but nowhere else you are going to see two locks at once, so it can be confusing.

Reply:
@gadmm is right: the `Unlock_exn` mechanism assumes that `Lock`/`Unlock` calls are lexically well-bracketted. In particular, an unrelated exception may very well unlock one of the debugger locks, which is clearly unexpected. The only reason this is not a bug is that when the debugger is loaded, systhread cannot be loaded, and hence locks are no-ops. 

Clearly, this is fragile, and I would prefer getting rid of these two locks, with a check that systhread is not loaded (and some comments explaining all that).